src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp

```diff
@@ -5573,6 +5573,309 @@ class StubGenerator: public StubCodeGenerator {
+    __ tbx(decH1, arrangement, codecH, 4, decH1);
+    __ tbx(decH2, arrangement, codecH, 4, decH2);
+    __ tbx(decH3, arrangement, codecH, 4, decH3);
+
+    // combine lower and higher
+    __ orr(decL0, arrangement, decL0, decH0);
+    __ orr(decL1, arrangement, decL1, decH1);
+    __ orr(decL2, arrangement, decL2, decH2);
+    __ orr(decL3, arrangement, decL3, decH3);
+
+    // check iilegal inputs, value larger than 63 (maximum of 6 bits)
+    __ cmhi(decH0, arrangement, decL0, v27);
+    __ cmhi(decH1, arrangement, decL1, v27);
+    __ cmhi(decH2, arrangement, decL2, v27);
+    __ cmhi(decH3, arrangement, decL3, v27);
+    __ orr(in0, arrangement, decH0, decH1);
+    __ orr(in1, arrangement, decH2, decH3);
+    __ orr(in2, arrangement, in0,   in1);
+    __ umaxv(in3, arrangement, in2);
+    __ umov(rscratch2, in3, __ B, 0);
+
+    // get the data to output
+    __ shl(out0,  arrangement, decL0, 2);
+    __ ushr(out1, arrangement, decL1, 4);
+    __ orr(out0,  arrangement, out0,  out1);
+    __ shl(out1,  arrangement, decL1, 4);
+    __ ushr(out2, arrangement, decL2, 2);
+    __ orr(out1,  arrangement, out1,  out2);
+    __ shl(out2,  arrangement, decL2, 6);
+    __ orr(out2,  arrangement, out2,  decL3);
+
+    __ cbz(rscratch2, NoIllegalData);
+
+    // handle illegal input
+    if (size == 16) {
+      Label ErrorInLowerHalf;
+      __ umov(rscratch1, in2, __ D, 0);
+      __ cbnz(rscratch1, ErrorInLowerHalf);
+
+      // illegal input is in higher half, store the lower half now.
+      __ st3(out0, out1, out2, __ T8B, __ post(dst, 24));
+
+      for (int i = 8; i < 15; i++) {
+        __ umov(rscratch2, in2, __ B, (u1) i);
+        __ cbnz(rscratch2, Exit);
+        __ umov(r10, out0, __ B, (u1) i);
+        __ umov(r11, out1, __ B, (u1) i);
+        __ umov(r12, out2, __ B, (u1) i);
+        __ strb(r10, __ post(dst, 1));
+        __ strb(r11, __ post(dst, 1));
+        __ strb(r12, __ post(dst, 1));
+      }
+      __ b(Exit);
+
+      __ BIND(ErrorInLowerHalf);
+    }
+
+    for (int i = 0; i < 7; i++) {
+        __ umov(rscratch2, in2, __ B, (u1) i);
+        __ cbnz(rscratch2, Exit);
+        __ umov(r10, out0, __ B, (u1) i);
+        __ umov(r11, out1, __ B, (u1) i);
+        __ umov(r12, out2, __ B, (u1) i);
+        __ strb(r10, __ post(dst, 1));
+        __ strb(r11, __ post(dst, 1));
+        __ strb(r12, __ post(dst, 1));
+    }
+    __ b(Exit);
+
+    __ BIND(NoIllegalData);
+    __ st3(out0, out1, out2, arrangement, __ post(dst, 3 * size));
+  }
+
+
+   /**
+   *  Arguments:
+   *
+   *  Input:
+   *  c_rarg0   - src_start
+   *  c_rarg1   - src_offset
+   *  c_rarg2   - src_length
+   *  c_rarg3   - dest_start
+   *  c_rarg4   - dest_offset
+   *  c_rarg5   - isURL
+   *
+   */
+  address generate_base64_decodeBlock() {
+
+    static const uint8_t fromBase64ForNoSIMD[256] = {
+      255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u,
```

Comment:
There seems to be no documentation of these magic tables of constants.

Reply:
We're either going to need a proper description of the algorithm here or a permalink to one.